File: CS/Reporting_how-to-create-a-label-report-at-runtime-t473792/LabelReportRuntime/LabelReportValuesHelper.cs
using DevExpress.XtraReports.Wizards.Labels;
using System.Collections.Generic;
using System.IO;
using System.Xml.Linq;

namespace dxWinFormsSample {
    public class LabelReportValuesHelper {
        private XDocumentLabelProductRepository repository;
        /// <summary>
        ///This class prodives methods that allow you to fetch label report type settings such as Label Products list and Product Details.
        ///Keep in mind that you can provide a custom XML with available Label Product values.
        ///To do so, pass a path to the custom XML to the LabelReportValuesHelper constructor.
        ///
        ///For the default Label Report Wizard to use this custom XML as well, specify the static DevExpress.Data.XtraReports.Labels.LabelWizardCustomization.ExternalLabelProductRepository property at the application startup.
        /// </summary>
        private const string defaultExternalLabelProductRepository = "Labels.xml";
        public LabelReportValuesHelper():this(defaultExternalLabelProductRepository){
        }
        public LabelReportValuesHelper(string xMLFilePath) {
            if (string.IsNullOrEmpty(xMLFilePath))
                xMLFilePath = defaultExternalLabelProductRepository;
            var type = typeof(LabelProductRepositoryFactory);
            var assembly = type.Assembly;
            Stream stream = assembly.GetManifestResourceStream(type, xMLFilePath);
            repository = new XDocumentLabelProductRepository(XDocument.Load(DevExpress.Utils.SafeXml.CreateReader(stream)));
        }
        public IEnumerable<LabelDetails> GetProductDetails(int productValue) {
            return repository.GetSortedProductDetails(productValue);
        }
        public IEnumerable<LabelProduct> GetLabelProducts() {
            return repository.GetSortedProducts();
        }
        public IEnumerable<PaperKindData> GetPaperKinds() {
            return repository.GetSortedPaperKinds();
        }
        public PaperKindData GetPaperKindData(int id) {
            return repository.GetPaperKindData(id);
        }
        public static float ConvertMillimeterToInch(float ld) {
            return DevExpress.XtraPrinting.GraphicsUnitConverter.Convert(ld, System.Drawing.GraphicsUnit.Millimeter, System.Drawing.GraphicsUnit.Inch);
        }
        public static float ConvertInchToMillimeter(float ld) {
            return DevExpress.XtraPrinting.GraphicsUnitConverter.Convert(ld, System.Drawing.GraphicsUnit.Inch, System.Drawing.GraphicsUnit.Millimeter);
        }
    }
}

File: VB/Reporting_how-to-create-a-label-report-at-runtime-t473792/LabelReportRuntime/LabelReportValuesHelper.vb
Imports DevExpress.XtraReports.Wizards.Labels
Imports System.Collections.Generic
Imports System.IO
Imports System.Xml.Linq

Namespace dxWinFormsSample
    Public Class LabelReportValuesHelper
        Private repository As XDocumentLabelProductRepository
        ''' <summary>
        '''This class prodives methods that allow you to fetch label report type settings such as Label Products list and Product Details.
        '''Keep in mind that you can provide a custom XML with available Label Product values.
        '''To do so, pass a path to the custom XML to the LabelReportValuesHelper constructor.
        '''
        '''For the default Label Report Wizard to use this custom XML as well, specify the static DevExpress.Data.XtraReports.Labels.LabelWizardCustomization.ExternalLabelProductRepository property at the application startup.
        ''' </summary>
        Private Const defaultExternalLabelProductRepository As String = "Labels.xml"
        Public Sub New()
            Me.New(defaultExternalLabelProductRepository)
        End Sub
        Public Sub New(ByVal xMLFilePath As String)
            If String.IsNullOrEmpty(xMLFilePath) Then
                xMLFilePath = defaultExternalLabelProductRepository
            End If
            Dim type = GetType(LabelProductRepositoryFactory)
            Dim assembly = type.Assembly
            Dim stream As Stream = assembly.GetManifestResourceStream(type, xMLFilePath)
            repository = New XDocumentLabelProductRepository(XDocument.Load(DevExpress.Utils.SafeXml.CreateReader(stream)))
        End Sub
        Public Function GetProductDetails(ByVal productValue As Integer) As IEnumerable(Of LabelDetails)
            Return repository.GetSortedProductDetails(productValue)
        End Function
        Public Function GetLabelProducts() As IEnumerable(Of LabelProduct)
            Return repository.GetSortedProducts()
        End Function
        Public Function GetPaperKinds() As IEnumerable(Of PaperKindData)
            Return repository.GetSortedPaperKinds()
        End Function
        Public Function GetPaperKindData(ByVal id As Integer) As PaperKindData
            Return repository.GetPaperKindData(id)
        End Function
        Public Shared Function ConvertMillimeterToInch(ByVal ld As Single) As Single
            Return DevExpress.XtraPrinting.GraphicsUnitConverter.Convert(ld, System.Drawing.GraphicsUnit.Millimeter, System.Drawing.GraphicsUnit.Inch)
        End Function
        Public Shared Function ConvertInchToMillimeter(ByVal ld As Single) As Single
            Return DevExpress.XtraPrinting.GraphicsUnitConverter.Convert(ld, System.Drawing.GraphicsUnit.Inch, System.Drawing.GraphicsUnit.Millimeter)
        End Function
    End Class
End Namespace
